sql/core/src/main/scala/org/apache/spark/sql/functions.scala

```diff
@@ -3900,6 +3900,61 @@ object functions {
    */
   def to_csv(e: Column): Column = to_csv(e, Map.empty[String, String].asJava)
 
+  /**
+   * A transform for timestamps and dates to partition data into years.
+   *
+   * @group partition_transforms
+   * @since 3.0.0
+   */
+  def years(e: Column): Column = withExpr { Years(e.expr) }
```

Comment:
I'm worried that these are a single letter away from existing function names. With one small typo, you either get your function to be unevaluable, or return a runtime exception.

Reply:
Can you think of a solution to this besides supporting `year(ts)` as though it were `years`? I'm concerned that would cause confusion for functions like `hour` that have different meanings.

Maybe we should catch these and throw more helpful exceptions instead?

Reply:
can we hide them by a namespace?
```
object partitionBy {
  def years: ...
  def months: ...
}
```

Reply:
So this would be `partitionBy.years`? I find `partitionBy` a little awkward. Is there a better name we could use? `partitioner.years` maybe?

Reply:
I'm not too stuck up on the name. What do you think of the idea in general? Other names I can think of:
 - partitioner
 - logical
 - transformer
 - partitioning
 - partition
could be like `partition.byYears`, `partition.byDays`...

Reply:
I think the names should match the SQL names, which are `years`, `months`, etc.

I think I like "partitioning" the best, since it qualifies the function. `partitioning.years`. Should I make this change?

Reply:
I'm fine with that. Let's discuss at the sync in 30 mins and gather feedback. Wouldn't want you to waste work.